File: pset5/speller/speller.cpp
/**
 * Implements a spell-checker.
 */

#include "speller.h"

#undef calculate
#undef getrusage

#include <iostream>
#include <iomanip>
using namespace std;

#define DEBUG

// default dictionary
#define DICTIONARY "dictionaries/large"

enum times {
    TIME_LOAD = 0, TIME_CHECK, TIME_SIZE, TIME_UNLOAD, TIME_TOTAL
};

#ifdef DEBUG
// debug info
void Speller::tagging(const char* NAME) {
    cout<<TAG<<" === "<<NAME<<" === "<<endl;
}
#endif

// constructor for debug
Speller::Speller() {
#ifdef DEBUG
    //tagging("Speller()");
#endif
};

// constructor with filenames
Speller::Speller(const char* dictName, char* textName) {
#ifdef DEBUG
    //tagging(TAG, "Speller(const char*, char*)");
#endif
    dictionary = dictName;
    text = textName;
};

// destructor  for debug
Speller::~Speller() {
#ifdef DEBUG
    //tagging("~Speller()");
#endif
};

// organization of resources to check words
int Speller::checking() {
#ifdef DEBUG
    //tagging(TAG, "checking");
#endif
    // class for measurement of time
    Stopwatch stopwatch;
    
    // class with template data structure
    Dictionary <Trie> dict;
    
    // load dictionary
    int res = loadDictionary(stopwatch, dict);
    if(res) {
        return res;
    }
    
    // try to open text
    FILE *fp = fopen(text, "rb");
    if (fp == NULL)
    {
        cout<<"Could not open "<<text<<"."<<endl;
        dict.unload();
        return 1;
    }
    
    // class for analyze words by letters
    WordsAudit* wordsAudit = new WordsAudit(stopwatch, dict, fp);
    
    // find words and report misspellings
    wordsAudit->findMisspellings();

    // check whether there was an error
    if (ferror(fp))
    {
        fclose(fp);
        cout<<"Error reading "<<text<<"."<<endl;
        dict.unload();
        return 1;
    }

    // close text
    fclose(fp);

    // determine dictionary's size
    unsigned int n = determineDictionarySize(stopwatch, dict);

    // unload dictionary
    res = unloaDictionary(stopwatch, dict);
    if(res) {
        return res;
    }
    
    // calculate total time
    stopwatch.calculate(TIME_TOTAL);

    // report benchmarks
    reportBenchmarks(stopwatch, wordsAudit->getMisspellings(), 
                             n, wordsAudit->getWords());

    // complete analysis
    delete wordsAudit;
    
    // that's all folks
    return 0;
}

// load dictionary
int Speller::loadDictionary(Stopwatch& stopwatch, Dictionary <Trie>& dict) {
#ifdef DEBUG
    //tagging(TAG, "loadDictionary");
#endif

    stopwatch.start();
    bool loaded = dict.load(dictionary);
    stopwatch.stop();

    // abort if dictionary not loaded
    if (!loaded)
    {
        cout<<"Could not load "<<dictionary<<"."<<endl;
        return 1;
    }

    // calculate time to load dictionary
    stopwatch.calculate(TIME_LOAD);
    
    return 0;
}

// determine dictionary's size
unsigned int Speller::determineDictionarySize(Stopwatch& stopwatch, Dictionary <Trie>& dict) {       
#ifdef DEBUG
    //tagging(TAG, "determineDictionarySize");
#endif

    stopwatch.start();
    unsigned int dictSize = dict.size();
    stopwatch.stop();

    // calculate time to determine dictionary's size
    stopwatch.calculate(TIME_SIZE);
    
    return dictSize;
}

// unload dictionary
int Speller::unloaDictionary(Stopwatch& stopwatch, Dictionary <Trie>& dict) { 
#ifdef DEBUG
    //tagging(TAG, "unloaDictionary");
#endif

    stopwatch.start();
    bool unloaded = dict.unload();
    stopwatch.stop();

    // abort if dictionary not unloaded
    if (!unloaded)
    {
        cout<<"Could not unload "<<dictionary<<"."<<endl;
        return 1;
    }

    // calculate time to unload dictionary
    stopwatch.calculate(TIME_UNLOAD);
    
    return 0;
}

// report benchmarks
void Speller::reportBenchmarks(Stopwatch& stopwatch, int misspellings, unsigned int& n, int words) {
#ifdef DEBUG
    //tagging(TAG, "reportBenchmarks");
#endif

    cout<<endl
        <<"WORDS MISSPELLED:     "<<misspellings<<endl
        <<"WORDS IN DICTIONARY:  "<<n<<endl
        <<"WORDS IN TEXT:        "<<words<<endl<<fixed<<setprecision( 2 )
        <<"TIME IN load:         "<<stopwatch.getResults(TIME_LOAD)<<endl
        <<"TIME IN check:        "<<stopwatch.getResults(TIME_CHECK)<<endl
        <<"TIME IN size:         "<<stopwatch.getResults(TIME_SIZE)<<endl
        <<"TIME IN unload:       "<<stopwatch.getResults(TIME_UNLOAD)<<endl
        <<"TIME IN TOTAL:        "<<stopwatch.getResults(TIME_TOTAL)<<endl<<endl;
}


int main(int argc, char *argv[])
{   
    // check for correct number of args
    if (argc != 2 && argc != 3)
    {
        cout<<"Usage: speller [dictionary] text"<<endl;
        return 1;
    }

    // determine dictionary to use
    const char* dictionary = (argc == 3) ? argv[1] : DICTIONARY;

    // try to open text
    char *text = (argc == 3) ? argv[2] : argv[1];

    // organization of spell-checking with definite dictionary and textfile
    Speller* speller = new Speller(dictionary, text);
    
    // organization of resources to check words
    speller->checking();

    // complete spell-checking
    delete speller;
    
    // that's all folks
    return 0;
}

File: pset7/speller/speller.go
/**
 * Implements a spell-checker.
 */

package main

import (
	"fmt"
	"os"
	"runtime"
	"time"
)

const (
	// maximum length for a word
	// (e.g., pneumonoultramicroscopicsilicovolcanoconiosis)
	LENGTH int = 45
	// default dictionary
	DICTIONARY string = "dictionaries/large"
	DEBUG      bool   = false
)

// enum times
const (
	TIME_LOAD  int = iota // 0
	TIME_CHECK            // 1
	TIME_SIZE             // 2
	TIME_TOTAL            // 3
)

var (
	workers = runtime.NumCPU()
	// name of dictionary file
	dictionary string
	// name of using textfile
	text string
	// pointer on using stopwatch
	stopwatch *Stopwatch
	// pointer on default datastructure
	dict *Dictionary
	// pointer on words analyzer
	wordsAudit *WordsAudit
)

func init() {
	//tagging("init")

	// class for measurement of time
	stopwatch = &Stopwatch{}
	// class with template data structure
	dict = &Dictionary{}
	// class for analyze words by letters
	wordsAudit = &WordsAudit{}
}

// organization of resources to check words
func checking() error {
	//tagging("checking")

	// load dictionary
	err := loadDictionary()
	if err != nil {
		return err
	}
	// check text file
	err = checkText()
	if err != nil {
		return err
	}
	// determine dictionary's size
	n := determineDictionarySize()

	// calculate total time
	stopwatch.calculate(TIME_TOTAL, 0.0)

	// report benchmarks
	reportBenchmarks(wordsAudit.misspellings, n, wordsAudit.words)

	// that's all folks
	return nil
}

// load dictionary
func loadDictionary() error {
	//tagging("loadDictionary")

	before := time.Now()
	err := dict.load()
	after := time.Now()

	// if dictionary not loaded
	if err != nil {
		return err
	}
	// calculate time to load dictionary
	stopwatch.calculate(TIME_LOAD, (after.Sub(before)).Seconds())

	return nil
}

// check text file
func checkText() error {
	//tagging("checkText")

	// check word's spelling
	before := time.Now()
	err := wordsAudit.check()
	after := time.Now()

	// if the check does not work
	if err != nil {
		return err
	}
	// update benchmark
	stopwatch.calculate(TIME_CHECK, (after.Sub(before)).Seconds())

	return nil
}

// report benchmarks
func reportBenchmarks(misspellings int, n uint, words int) {
	//tagging("reportBenchmarks")

	fmt.Print("\n",
		"WORDS MISSPELLED:     ", misspellings, "\n",
		"WORDS IN DICTIONARY:  ", n, "\n",
		"WORDS IN TEXT:        ", words, "\n")
	fmt.Printf("TIME IN load:         %.2f\n", stopwatch.getResults(TIME_LOAD))
	fmt.Printf("TIME IN check:        %.2f\n", stopwatch.getResults(TIME_CHECK))
	fmt.Printf("TIME IN size:         %.2f\n", stopwatch.getResults(TIME_SIZE))
	fmt.Printf("TIME IN TOTAL:        %.2f\n\n", stopwatch.getResults(TIME_TOTAL))
}

func main() {

	lenArgs := len(os.Args)
	// check for correct number of args
	if lenArgs != 2 && lenArgs != 3 {
		fmt.Print("Usage: speller [dictionary] text\n")
		return
	}
	if lenArgs == 3 {
		// determine dictionary to use
		dictionary = os.Args[1]
		// textfile to open text
		text = os.Args[2]

	} else {
		dictionary = DICTIONARY
		text = os.Args[1]
	}
	// organization of resources to check words
	checking()

	// that's all folks
	return
}
